### ISSUE TITLE
Stack trace isn't being de-obfuscated for andorid. R8 code shrinker.

### ISSUE THREAD

**Issue by CarsonRedeye** [getsentry/rust-proguard#6]
Mapping files are uploading correctly via the sentry-gradle-plugin. I can see them under Debug Files in sentry settings. However I can only see a stack trace like this:

`java.lang.NumberFormatException: For input string: "bucket_3017"
    at java.lang.Long.parseLong(Long.java:594)
    at java.lang.Long.parseLong(Long.java:636)
    at c.a.a.e.b0.g.b.a(SourceFile:96)
    at co.redacted.presentation.search.qrcode.b.a(SourceFile:26)
    at co.redacted.presentation.search.qrcode.b.a(SourceFile:3)
    at co.redacted.presentation.search.qrcode.b$g.a(SourceFile:2)
    at co.redacted.presentation.search.qrcode.b$g.a(SourceFile:1)
    at g.b.b0.d.f.onSuccess(SourceFile:2)
    at g.b.b0.e.f.d$a.onSuccess(SourceFile:1)
    at g.b.b0.e.f.e$a.onSuccess(SourceFile:2)
    at g.b.b0.e.f.p$a.onSuccess(SourceFile:1)
    at g.b.b0.e.e.z$a.onComplete(SourceFile:6)
    at n.v.a.a$a.onComplete(SourceFile:2)
    at n.v.a.c.b(SourceFile:9)
    at g.b.m.a(SourceFile:54)
    at n.v.a.a.b(SourceFile:1)
    at g.b.m.a(SourceFile:54)
    at g.b.b0.e.e.z.b(SourceFile:1)
    at g.b.s.a(SourceFile:32)
    at g.b.b0.e.f.p$a.run(SourceFile:1)
    at g.b.r$a.run(SourceFile:2)
    at g.b.b0.g.k.run(SourceFile:2)
    at g.b.b0.g.k.call(SourceFile:1)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    at java.lang.Thread.run(Thread.java:764)`

If I download that same mapping file from sentry and use Analyze Stack Track in Android studio, and apply Proguard Unscrambler it works correctly. My suspicion is that changes to the mapping file with new versions of Android Studio/R8 shrinker are causing sentry.io to fail with it's deobfuscation.

I'm using 'io.sentry:sentry-android-gradle-plugin:1.7.21' and 'io.sentry:sentry-android:1.7.21'

**Comment by TobiasReich** [comment]
So are there any news?

**Comment by marandaneto** [comment]
@TobiasReich not yet, sorry for that, I'll try myself over the weekend.
PR is also welcomed :) thanks and stay tuned.

**Comment by marandaneto** [comment]
ok here is how I've reproduced it and the screenshots that show the problem.

Symbolicated by Sentry (pay attention at line 1 and 2):
<img width="901" alt="Screenshot 2020-03-29 at 11 16 41" src="https://user-images.githubusercontent.com/5731772/77845561-18a6f880-71b0-11ea-88e3-8e6a5ad84417.png">

Raw stack trace:
<img width="910" alt="Screenshot 2020-03-29 at 11 20 39" src="https://user-images.githubusercontent.com/5731772/77845577-3ecc9880-71b0-11ea-94e0-7b10ec33a408.png">

Symbolicated by Retrace (proguard):
<img width="937" alt="Screenshot 2020-03-29 at 11 21 23" src="https://user-images.githubusercontent.com/5731772/77845582-4ee47800-71b0-11ea-8b30-d41fa7765760.png">

Mapping file:
[mapping.txt](https://github.com/getsentry/rust-proguard/files/4398535/mapping.txt)

Raw stack trace file:
[raw stacktrace.txt](https://github.com/getsentry/rust-proguard/files/4398537/raw.stacktrace.txt)

So I believe this problem happens when there's a method reference and/or Lambdas on Java 8, I've not tested on Kotlin btw.

eg:
`button.setOnClickListener(this::buttonClicked);`

The source of the problem is line numbers, When symbolicated by Sentry, it shows:
doWork **at line 2**
onClick **at line 2**
_both are wrong._

When symbolicated by Retrace, it shows:
MyBean **at line 16**
MyActivity **at line 29**
_which matches the source code._

Curious that Retrace shows a 3rd unsymbolicated line _d.c.a.a.onClick_ and Sentry successfully symbolicated to the generated Lambda name.

I've used AGP 3.6.1 so minifying is powered by R8.
I've used Proguard/Retrace 6.2.2.

**Comment by marandaneto** [comment]
Same tests, but using Proguard instead of R8.

Retrace does it correctly:
<img width="575" alt="Screenshot 2020-03-29 at 12 06 14" src="https://user-images.githubusercontent.com/5731772/77846349-cf59a780-71b5-11ea-9cc7-3ebe4ea50bed.png">

Sentry still mistakes the line numbers:
<img width="916" alt="Screenshot 2020-03-29 at 12 08 44" src="https://user-images.githubusercontent.com/5731772/77846408-1f386e80-71b6-11ea-862d-a5131c9fefba.png">

Mappings file:
[mapping.txt](https://github.com/getsentry/rust-proguard/files/4398602/mapping.txt)

Raw stack trace:
[raw stacktrace.txt](https://github.com/getsentry/rust-proguard/files/4398606/raw.stacktrace.txt)

**Comment by Swatinem** [comment]
So here is a short summary from when I was looking into this.
#10 has a few testcases, from both #5 and the two mapping files you linked here.
I also linked to https://www.guardsquare.com/en/products/proguard/manual/retrace there, which has a good explanation and also an example.

Tracing this code through symbolic as well I found out that it only does class/method mapping here:
https://github.com/getsentry/symbolic/blob/fd22b1a8f970d94832feae5c9971302c12778696/proguard/src/lib.rs#L50
https://github.com/getsentry/sentry/blob/d46213a3cb4fea1653724ac58bcd78e32a3dda05/src/sentry/lang/java/plugin.py#L110

What actually needs to be done here is two things:
* do actual line-number re-mapping, similar to how sourcemaps in JS work
* synthesize stackframes for inlined methods, and detect those based on the mapping file

so, tldr: This would need deeper changes to how sentry/symbolic/proguard work together.

**Comment by marandaneto** [comment]
@Swatinem thanks for the summary.

actually https://github.com/getsentry/sentry-android-gradle-plugin does add extra rules for that:
`-keepattributes LineNumberTable,SourceFile`

does it not help at all?

**Comment by Swatinem** [comment]
Doesn’t seem like it does. The line number in your stack trace is a "synthetic" one, as explained in the retrace docs. They need to be remapped based on the mapping file.

**Comment by Swatinem** [comment]
Merging the PR actually auto-closed this. Not sure if we should re-open, or rather move this into the sentry repo, as the code that actually processes the stacktrace lives there.

**Comment by marandaneto** [comment]
@Swatinem I'd keep it open, but moving sounds a good idea.

**Comment by TobiasReich** [comment]
So is there a new release any time soon?

**Comment by Swatinem** [comment]
I’m currently in the process of wrapping this up. You can take a peek at the finished symbolication here: https://github.com/getsentry/sentry/pull/19012#issue-423190113

**Comment by TobiasReich** [comment]
Good morning, is there an ETA for the fix already?

**Comment by marandaneto** [comment]
It's fixed and deployed on sentry.io

Please let us know if you find any issues, thanks.
